src/fs/data_compaction.cc

```diff
@@ -0,0 +1,397 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+#include "fs/bitwise.hh"
+#include "fs/cluster.hh"
+#include "fs/data_compaction.hh"
+#include "fs/device_reader.hh"
+#include "fs/inode_info.hh"
+#include "fs/metadata_log_operations/write.hh"
+#include "fs/units.hh"
+#include "seastar/core/do_with.hh"
+#include "seastar/core/future-util.hh"
+#include "seastar/core/future.hh"
+#include "seastar/core/temporary_buffer.hh"
+
+#include <algorithm>
+#include <boost/iterator/counting_iterator.hpp>
+#include <exception>
+#include <optional>
+#include <stdexcept>
+#include <variant>
+
+namespace seastar::fs {
+
+namespace {
+logger mlogger("fs_compaction");
+} // namespace
+
+future<> data_compaction::compact() {
+    mlogger.debug(">>>>  Started compaction  <<<<");
+    mlogger.debug("source clusters: {}", _compacted_cluster_ids);
+
+    auto allocate_compacted_data_vec = [&](disk_offset_t disk_offset,
+            const data_cluster_contents_info::cluster_data_vec& data) {
+        compacted_data_vec vec {
+            ._data = temporary_buffer<uint8_t>::aligned(_metadata_log._alignment,
+                    round_up_to_multiple_of_power_of_2(data.data_range.size(), _metadata_log._alignment)),
+            ._inode_id = data.data_owner,
+            ._file_offset = data.data_range.beg,
+            ._prev_disk_offset = disk_offset,
+            ._post_disk_offset = std::nullopt
+        };
+        vec._data.trim(data.data_range.size());
+        return vec;
+    };
+    std::sort(_compacted_cluster_ids.begin(), _compacted_cluster_ids.end()); // TODO: verify if this is needed here or
+                                                                             //       is it already done by caller
```

Comment:
That TODO looks like it should be resolved before merging this.

Reply:
Here I meant that caller (something that decides what to compact and when to compact) may pass those clusters in some order. So if caller sorted those clusters before calling compaction then we wouldn't need that sort here but it depends on future implementations. By the way, compaction works even without sorting those clusters but I thought that sorting would be beneficial because of sequential reads. So we could actually remove that sort for now and decide whether we want to sort clusters in the future implementation of the caller.